feat(tests/ioworker): add option for nlb

karlowich opened the pull request:
The nlb option allows for testing larger payloads

karlowich committed:
fix(tests/ioworker): tidy up

- Change debug messages to properly reflect err
- Fill entire read buffer with zeros instead of only 512 bytes

Signed-off-by: Karl Bonde Torp <k.torp@samsung.com>

karlowich committed:
fix(tests/ioworker): zero the full read buffer

The read buffer is allocated and compared over range.nbytes, but was only
zero-filled for io.nbytes -- a single I/O's worth. Fill the whole buffer so
its initial contents are well-defined across the entire range.

Also correct the buf_alloc debug labels (data/buf -> wbuf/rbuf).

Signed-off-by: Karl Bonde Torp <k.torp@samsung.com>

karlowich committed:
feat(tests/ioworker): add option for nlb

Signed-off-by: Karl Bonde Torp <k.torp@samsung.com>

karlowich committed:
feat(tests/ioworker): add option for nlb

Signed-off-by: Karl Bonde Torp <k.torp@samsung.com>

karlowich committed:
feat(cijoe/ioworker): expand parameterization

Signed-off-by: Karl Bonde Torp <k.torp@samsung.com>

karlowich committed:
feat(cijoe/ioworker): expand parameterization

Signed-off-by: Karl Bonde Torp <k.torp@samsung.com>

safl commented:
Rebase this before running tests again, there is an issue with Windows causing the job to run for about 5 hours.

karlowich commented:
Found 2 issues:

1. **Async path not advancing slba**: `_submit()` never updates `work->cur.slba` after vector loop, while `_submit_sync` does at line ~187. This causes subsequent I/Os to use stale SLBA values leading to PRP misalignment errors. Fix: add `work->cur.slba += work->io.naddr;` after the vector loop.

2. **Buffer offset mismatch** (line ~144): Non-vectored path uses `work->geo->lba_nbytes` instead of `work->io.nbytes`. May cause incorrect offsets when nlb > 0.

---
*Review by Karlowich OpenCode - automated review, may contain errors*

karlowich commented on tests/ioworker.c:
```diff
@@ -112,10 +114,10 @@ _submit(struct iowork *work, struct ioworker *worker, struct xnvme_cmd_ctx *ctx)
 				work->cur.data + work->cur.slba * work->geo->lba_nbytes;
-			worker->vec[j].iov_len = work->geo->lba_nbytes;
+			worker->vec[j].iov_len = (work->io.nlb + 1) * work->geo->lba_nbytes;
 		}
```
Bug: work->cur.slba is not updated after this loop. The sync path (_submit_sync) correctly updates it at line 187 with work->cur.slba += work->io.nlb + 1, but this async path has no such update. This causes subsequent I/Os to use stale SLBA values, leading to PRP misalignment. Fix: add work->cur.slba += work->io.naddr; after this loop.

karlowich commented on tests/ioworker.c:
```diff
@@ -139,7 +141,7 @@ _submit(struct iowork *work, struct ioworker *worker, struct xnvme_cmd_ctx *ctx)
 	} else {
 		work->cur.slba = ctx->cmd.nvm.slba;
-		void *dbuf = work->cur.data + work->cur.slba * work->io.nbytes;
+		void *dbuf = work->cur.data + work->cur.slba * work->geo->lba_nbytes;
```
Potential bug: Original code used work->io.nbytes here. Current code uses work->geo->lba_nbytes which may cause incorrect buffer offsets when nlb > 0.

karlowich committed:
fix(tests/ioworker): drop work->cur, use ctx->cmd.nvm.slba for SLBA tracking

Replace the single shared work->cur cursor (data pointer + slba) with a
single work->buf and per-command SLBA tracking via ctx->cmd.nvm.slba,
which the backends preserve until completion.

For the async path each worker owns one command context and a disjoint
slice of the range: worker i starts at 'range.slba + i * io.naddr' and, on
completion, advances its SLBA by 'nworkers * io.naddr' so the workers
interleave without overlapping. A worker stops once its next I/O would
fall outside the range. Previously a worker advanced by a single io.naddr,
so the workers trampled each other's LBAs and left the upper part of the
range unwritten, failing the wbuf/rbuf comparison.

Signed-off-by: Karl Bonde Torp <k.torp@samsung.com>

karlowich committed:
fix(be/linux/libaio): keep iocb in local storage, preserve ctx->cmd

The libaio backend aliased 'struct iocb' onto '&ctx->cmd' and let
io_prep_*()/io_submit() overwrite it, clobbering the NVMe command. Unlike
the io_uring backends -- which build their SQE in the ring's own memory and
leave 'ctx->cmd' intact -- this left the command unusable after submission.

The iocb is only read by the kernel during io_submit(), so hold it in a
local variable instead. 'ctx->cmd' now survives until completion, matching
the other backends, so callers can rely on it (e.g. to track SLBA across
re-submissions).

Signed-off-by: Karl Bonde Torp <k.torp@samsung.com>

safl commented on tests/ioworker.c:
```diff
@@ -298,9 +298,8 @@ iowork_from_cli(struct xnvme_cli *cli, struct iowork *work)
-	work->io.naddr = work->vec_cnt;
-	work->io.nlb = work->io.naddr - 1;
+	work->io.nbytes = work->vec_cnt * (work->io.nlb + 1) * work->geo->lba_nbytes;
+	work->io.naddr = work->vec_cnt * (work->io.nlb + 1);
```
Guard both failures right after this line:

```c
if (!work->io.naddr || work->range.naddr % work->io.naddr) {
    XNVME_DEBUG("FAILED: io.naddr must divide range.naddr");
    return -EINVAL;
}
```

Rejects `nio=0` (io.naddr > range.naddr) and non-divisible `--nlb` (tail-unwritten diff EIO). Also forecloses the `io.nbytes` overrun and 16-bit `nlb` truncation, since both need `io.naddr > range.naddr` first.

karlowich replied:
Great catch!
It's fixed now.

naddinadja commented on cijoe/tests/logic/test_xnvme_tests_ioworker.py:
```diff
@@ -50,18 +78,54 @@ def test_verify_sync_iovec(cijoe, device, be_opts, cli_args):
     else:
         assert not err
 
+    # small nlb: sub-page segments require an SGL, rejected on nosgl
+    err, _ = cijoe.run(
+        f"xnvme_tests_ioworker verify-sync {cli_args} --nlb 1 --vec-cnt 4"
+    )
+
+    if be_opts["be"] == "spdk" and "nosgl" in device["labels"]:
+        assert err
+    else:
+        assert not err
```
A question for this: is there something inherent in our code that MUST fail if "nosgl" is in labels, or is it just that "this is not available and will therefore fail"? Because if it's the latter, then I think it would be better with a `pytest.skip()`

(there are 4 instances of this in this file, i only comment once)

karlowich replied:
Yes, if the device doesn't support SGLs then we can't handle iovecs with segments smaller than the host page size. If the segments are matching the host page size, then we can use PRP lists instead of SGLs on these nosgl devices.

safl commented on tests/ioworker.c:
```diff
@@ -298,9 +298,13 @@ iowork_from_cli(struct xnvme_cli *cli, struct iowork *work)
+	work->io.nbytes = work->vec_cnt * (work->io.nlb + 1) * work->geo->lba_nbytes;
+	work->io.naddr = work->vec_cnt * (work->io.nlb + 1);
+
+	if (!work->io.naddr || work->range.naddr % work->io.naddr) {
```
This runs before `work->range.naddr` is assigned a few lines below, so it compares against 0 (the buffer is zeroed at function entry) and `0 % io.naddr == 0`, so the check is dead. `--nlb 2` on a 4K-LBA device (`io.naddr=3`, `range.naddr=4096`) passes here and then fails at `xnvme_buf_diff`, which is the case this guard was meant to catch. Move the block below the `work->range.naddr = ...` assignment.

karlowich replied:
Now it should be correct. The only thing changed was that block

safl commented on tests/ioworker.c:
```diff
@@ -232,16 +229,18 @@ on_completion(struct xnvme_cmd_ctx *ctx, void *cb_arg)
-	if (work->cur.slba >= work->range.elba) {
+	// Workers interleave, each striding by 'nworkers * io.naddr'; stop this worker when its
+	// next I/O would fall outside the range.
+	if (ctx->cmd.nvm.slba + (work->nworkers + 1) * work->io.naddr > work->range.naddr) {
```
Uses `range.naddr` as if it were the exclusive end LBA; only correct because `range.slba` is hardcoded to 0. Using `range.slba + range.naddr` on the right-hand side keeps this robust if a subcommand ever starts the range mid-device.

karlowich replied:
fixed

safl commented on tests/ioworker.c:
```diff
@@ -298,23 +298,27 @@ iowork_from_cli(struct xnvme_cli *cli, struct iowork *work)
 	work->range.nbytes = 1 << 24;
 	work->range.naddr = work->range.nbytes / work->geo->lba_nbytes;
 	work->range.slba = 0;
 	work->range.elba = work->range.slba + work->range.naddr - 1;
```
After the SLBA-tracking rework nothing reads `range.elba` or `range.nlb` except the `iowork_pp` printout of `elba`; both assignments (and the fields themselves) could go.

karlowich replied:
fixed
